---
title: "Supplemental"
output:
  html_document: default
  pdf_document: default
---

# Setup

This document will make Supplemental Figures which are not made during the creation of main text figures (so not Supp Fig 3, 7, and 11). I don't do as much rearranging as for the main text figures (i.e. with cowplot -- though I still use the library because it looks nice!) so this is a little messier but makes the data reproducible.

```{r setup, message=FALSE, warning=FALSE, include=FALSE}
##Libraries
library(readr)
library(ggplot2)
library(data.table)
library(cowplot)
library(dplyr)

## My default Graphpad-like theme
theme<-theme_cowplot(font_size=10,line_size = 1) + theme(panel.background=element_rect(fill="white"),line=element_line(size=1,colour="black",lineend="round"),axis.line=element_line(size=1),text=element_text(size=10,face="bold",colour="black"),axis.text=element_text(colour="black"),axis.ticks=element_line(size=1,colour="black"),axis.ticks.length=unit(.1,"cm"),strip.background=element_blank(),legend.position="null",axis.text.x=element_text(angle=45,hjust=1))

## Empty plot to squish the figure together
empty <- ggplot() + geom_point(aes(1,1), colour="white") + theme(axis.ticks=element_blank(), panel.background=element_blank(), axis.text.x=element_blank(), axis.text.y=element_blank(), axis.title.x=element_blank(), axis.title.y=element_blank(),axis.line=element_blank())

##Calculate CI function for poisson distribution:
calc.ci<-function(n,lambda) {
  return(1.96*sqrt(lambda))
}

## ALU, LINE1, SVA, ALL colour palette
cPalette <- c("#7DA1D4","#BEDE85","#DCB070","#EAAAB0")
```

# Mosaic Supplemental Figures

This code performs the analysis of low variant allele frequency (VAF; a.k.a putative mosaic) MEIs in Supp Fig 2. The figures here just print out as is, need to actually arrange into final format and add legends in Adobe Illustrator.

## Analysis of Genotypes - 2a

Raw data here (`genodepth.di.txt`) was generated by calculating allelic depth with a internally modified version of MELT.

```{r Load Mosaic, fig.height=4, fig.width=8}

genodepth <- fread("RawData/genodepth.di.txt")
setnames(genodepth,names(genodepth),c("INS","CHR","POS","GT","AC","DP","AD","SAMP"))
genodepth[,SAMP:=as.character(SAMP)]

genodepth[,AP:=AD/DP]
genodepth[,AP.cut:=cut(AP,breaks=50)]

## This data.table is a manually entered set of coordinates for the low VAF variants. Only need position since no two sites in the genome from different chromosomes have the same coordinate. Proband and Parent IDs are Internal references not used anywhere else.
mosaic.sites <- data.table(POS =c(74349692,200586931,79910459,125752471,40407986,55870411,71004747,101378585,101802070,11416279,95523918,70188151,91145646,47696522,30258814),
                           PRO=c("20818","13033","08461","27847","22252","25672","20644","02758","23167","04729","02452","29248","06364","10333","11653"),
                           PARENT=c("20819","13034","08463","27848","22253","25674","20645","02760","23168","04730","02454","29249","06366","10334","11655"))


mosaic.sites<-merge(mosaic.sites,genodepth,by.x=c("POS","PRO"),by.y=c("POS","SAMP"),all.y=F)
mosaic.sites<-merge(mosaic.sites,genodepth,by.x=c("POS","PARENT"),by.y=c("POS","SAMP"),all.y=F,suffixes=c(".proband",".parent"))

mean.acOne <- mean(genodepth[DP >= 20 & AC == 1, AP])
mean.acTwo <- mean(genodepth[DP >= 20 & AC == 2, AP])

mean.acOne
mean.acTwo

genodepth[,dummy:=1]
mosaic.sites[,dummy:=1]

attach <- function(cut,allele) {
  x<-nrow(genodepth[DP >= 10 & AC == allele & AP.cut == cut])
  return(x)
}

plottable <- genodepth[DP >= 10,sum(dummy),by=c("AP.cut","AC")]
plottable <- data.table(AP.cut=rep(unique(plottable[,AP.cut]),3),AC=rep(c(0:2),50))
plottable[,V1:=mapply(attach,AP.cut,AC)]
plottable[,V1:=mapply(ifelse,V1 > 4000,4000,V1)]

attach <- function(cut,allele) {
  x<-nrow(mosaic.sites[DP.parent >= 10 & AC.parent == allele & AP.cut.parent == cut])
  return(x)
}

plottable.test <- mosaic.sites[DP.parent >= 10,sum(dummy),by=c("AP.cut.parent","AC.parent")]
plottable.test <- data.table(AP.cut=rep(unique(plottable.test[,AP.cut.parent]),3),AC=rep(c(0:2),50))
plottable.test[,V1:=mapply(attach,AP.cut,AC)]
plottable.test[,V1:=mapply(ifelse,V1 > 4000,4000,V1)]

theme.this.chunk <- theme + theme(panel.grid.major=element_line(colour="grey",size=0.5))

all <- ggplot(plottable,aes(AP.cut,V1,colour=as.factor(AC),group=as.factor(AC))) + geom_line(size=1) + geom_col(inherit.aes = F, data = plottable.test,aes(AP.cut,V1)) + scale_x_discrete(labels = c((0:5) / 5), breaks = levels(unique(plottable.test[,AP.cut]))[c(1,(0:5)*10)], name = "Allelic Proportion") + ylab ("Total Genotypes") + theme.this.chunk
all
```


## Single Parent vs Proband - 2b

Have to add annotation for p.values in Adobe Illustrator, but p.values are calculated here via wilcox.test.

```{r fig.height=4, fig.width=4, message=FALSE, warning=FALSE}

boxplot <- data.table(ap=c(genodepth[AC==1,AP],mosaic.sites[,AP.parent],mosaic.sites[,AP.proband]),sample=c(rep("All Hets",nrow(genodepth[AC==1])),rep("Parents",nrow(mosaic.sites)),rep("Probands",nrow(mosaic.sites))))
boxplot[,sample:=factor(sample,levels=c("All Hets","Probands","Parents"))]

wilcox.test(ap~sample,data=boxplot[sample=="All Hets" | sample == "Probands"])
wilcox.test(ap~sample,data=boxplot[sample=="All Hets" | sample == "Parents"])

median(boxplot[sample=="All Hets",ap]) - median(boxplot[sample=="Parents",ap]) 
median(boxplot[sample=="All Hets",ap]) - median(boxplot[sample=="Probands",ap]) 

plot<-ggplot(boxplot,aes(sample,ap)) + geom_boxplot() + scale_y_continuous(name ="Allelic Proportion", limits = c(0,1.1), breaks=c(0:6/5)) + xlab("") + theme.this.chunk + theme(panel.grid.major.x=element_blank())
plot

```

### Per Site: - 2c

This is just plotting the mean allele depth at all individuals against just the parents with the identified low VAF variant.

```{r Compare to Parents, fig.height=3, fig.width=4, message=FALSE, warning=FALSE}
test<-genodepth[AC==0,list(mean(AP),sd(AP)),by="INS"]
setnames(test,names(test),c("INS","mean","sd"))
test[,sd.low:=mean-sd]
test[,sd.high:=mean+sd]
test[,Mosaic:=INS %in% mosaic.sites[,INS.proband]]
setkey(test,Mosaic)
test[,INS:=factor(INS,levels=INS)]
mosaic.sites[,coord:=paste0(CHR.parent,":",POS)]
test <- merge(test,mosaic.sites[,c("INS.proband","coord")],by.x="INS",by.y="INS.proband")

ggplot(test[Mosaic==T],aes(coord,mean)) + geom_point(color="black") + geom_errorbar(aes(ymin=sd.low,ymax=sd.high),width=0) + geom_point(inherit.aes=F,data=mosaic.sites,aes(coord,AP.parent),colour="red") + xlab("Insertion") + ylab("Alleleic Proportion") + theme.this.chunk + theme(panel.grid.major.y=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank()) + coord_flip()
```

# Supplemental Figure 4

See main paper methods for how the hardcoded numbers were generated. `\<MEI\>_random.txt` was generated by randomly downsampling the DDD parents to the sample size of the 1KGP using a command like: `vcftools --vcf <MEI>.vcf.gz --max-indv 2453 --non-ref-ac 1`. The 100 random simulations performed for the manuscript are what is loaded by this code chunk.

```{r Supplemental Figure 4, fig.height=4, fig.width=8}
##Sup Fig 2
ALU_random <- fread("RawData/ALU_random.txt")
setnames(ALU_random,"V1","count")
ALU_random[,mei:="Alu"]
LINE1_random <- fread("RawData/LINE1_random.txt")
setnames(LINE1_random,"V1","count")
LINE1_random[,mei:="LINE-1"]
SVA_random <- fread("RawData/SVA_random.txt")
setnames(SVA_random,"V1","count")
SVA_random[,mei:="SVA"]
  
##Get p-values from the random distribution:
##Numbers subtracted by the mean are based on actual 1KGP values - do not change
ALU.p <- 2*pnorm(-abs((318-mean(ALU_random[,count]))/sd(ALU_random[,count])))
LINE1.p <- 2*pnorm(-abs((81-mean(LINE1_random[,count]))/sd(LINE1_random[,count])))
SVA.p <- 2*pnorm(-abs((26-mean(SVA_random[,count]))/sd(SVA_random[,count])))

random <- bind_rows(ALU_random,LINE1_random,SVA_random)

random.line <- data.frame(mei=c("Alu","LINE-1","SVA"),
                          line=c(318,81,26),
                          line.txt=c(310,90,45),
                          text=c(paste("p=",format(ALU.p,digits=2)),
                                 paste("p=",format(LINE1.p,digits=2)),
                                 paste("p=",format(SVA.p,digits=2))),
                          height=c(20,40,110),
                          height.text=c(27,47,118))

theme.this.chunk <- theme %+replace% theme(legend.position=c(0.9,0.8))
plot<-ggplot(random,aes(count,fill=mei)) + scale_fill_manual(values=cPalette,guide=guide_legend(title="MEI"),labels=c(expression(bolditalic("Alu")),"L1","SVA")) + geom_histogram(binwidth=1) + xlab("Total MEI Sites") + ylab("# of Tests") + scale_y_continuous(expand=c(0,0)) + geom_linerange(data = random.line, aes(line,ymin=0,ymax=height),colour="red",size=1) + geom_text(data=random.line,aes(x=line.txt,y=height.text,label=text),size=4,colour="red") + scale_x_continuous(breaks=c(0,50,100,150,200,250,300,350)) + theme.this.chunk
plot
```

# PPG Supplemental Figures:

## Supplemental Figure 5

I have included in `RawData` a version of (Supplemental Table S5)[https://doi.org/10.1371/journal.pcbi.1005567.s005] from Zhang et. al. as the file `gerstein.txt`. This just includes each genes allele count in their data with ENSG id.

```{r Supplemental Figure 5}

gerstein <- fread("RawData/gerstein.txt")
setnames(gerstein,names(gerstein),c("geneid","N"))

pseudo_calls <- fread("../Figure1/RawData/Retrogenes.di.txt")
RETROGENE <- data.table(geneid=pseudo_calls[,GENE],N=rowSums(pseudo_calls[,5:ncol(pseudo_calls)]))

##This makes figure A
counts<-merge(gerstein,RETROGENE,by="geneid",all.x=T,all.y=T)
setnames(counts,c("N.x","N.y"),c("gerstein","gardner"))
counts<-counts[gerstein != 0][gardner!=0]
counts[,gerstein.log10:=log10(gerstein)]
counts[,gardner.log10:=log10(gardner)]
r2<-lm(gerstein.log10~gardner.log10,data=counts)
print(summary(r2))
plot<-ggplot(counts,aes(gerstein.log10,gardner.log10)) + geom_smooth(method=lm, colour="red") + geom_point() + scale_x_continuous(breaks=c(0,1,2,3),labels=c(1,10,100,1000)) + scale_y_continuous(breaks=c(0,1,2,3,4),labels=c(1,10,100,1000,10000)) + xlab("Gene allele counts in Zhang et. al.") + ylab("Gene allele counts in this study")
plot

##This makes figure B
counts<-merge(gerstein,RETROGENE,by="geneid",all.x=T,all.y=T)
setnames(counts,c("N.x","N.y"),c("gerstein","gardner"))
##Makes all NA into 0
counts[is.na(counts)] <- 0
counts[,is.known:=gerstein>0]

total_indv<-ncol(pseudo_calls) - 4 ## minus 4 because of info columns
counts[,af:=gardner/total_indv]
counts[,af.cut:=cut(af,breaks= c(0.00001,0.0001,0.001,0.01,0.1,1))]

counts<-counts[gardner>0]

known<-data.table(table(counts[is.known==T,af.cut]))
unknown<-data.table(table(counts[is.known==F,af.cut]))

ac<-data.table(ac=c(known[,V1],unknown[,V1]),N=c(known[,N],unknown[,N]),is.ps=c(rep(TRUE,5),rep(FALSE,5)))
ac<-data.frame(ac)
ac$ac2<-factor(ac$ac,levels=c("(1e-05,0.0001]","(0.0001,0.001]","(0.001,0.01]","(0.01,0.1]","(0.1,1]"))
ac<-data.table(ac)

theme.this.chunk<-theme %+replace% theme(legend.position=c(0.7,0.8))
plot<-ggplot(ac,aes(ac2,N,fill=is.ps)) + geom_col() + scale_x_discrete(labels=c("0.00001","0.0001","0.001","0.01","0.1")) + scale_y_continuous(expand=c(0,0)) +  xlab("Allele Frequency Bins") + ylab("Total Sites") + theme.this.chunk
plot
```

## Supplemental Figure 6

See main text methods for where `GTEx.expr.txt` came from. The file here is just simplified to be the mean/sd TPM across all individuals in GTEx for all ENSG ids so can link to the Retrogene data -- loaded in the above chunks.

`EXacPLI.txt` is just a subset of Supplemental Table 13 from [Lek. et. al.](https://www.nature.com/articles/nature19057) that includes gene names and pLI scores so we remove genes that weren't included in the PPG analysis for this paper but are annotated by GTEx.

```{r Supplemental Figure 6, fig.height=10, fig.width=8}

expression <- fread("RawData/GTEx.expr.txt")
setnames(expression,names(expression),c("gene.id","gene.name","tissue","mean","sd","n"))

expression[,sd.high:=(mean+sd)]
expression[,sd.low:=(mean-sd)]

ExacPLI <- fread("RawData/ExacPLI.txt")
setnames(ExacPLI,names(ExacPLI),c("gene.id","gene.name","chr","n.exons","start","end","pLI","pRec","pNull"))

## These values come from the Gerstein_Comparison.R script
expression<-merge(expression,RETROGENE,by.x="gene.id",by.y="geneid",all.x=T,all.y=F)
##Add pLI values
expression<-merge(expression,ExacPLI,by="gene.name",all.x=T,all.y=F)
## Remove genes/values that have no expression value and don't have a pLI score
expression<-expression[mean!=0]
expression<-expression[is.na(pLI) == F]
## Create boolean is.ps and numeric (for wilcoxon rank sum)
expression[,is.ps:=!is.na(N)]

convert.ps <- function(x) {
  if (x == T) {
    0
  } else {
    1
  }
}

## This just does the chi^2 test of high pLI donors vs overall high pLI genes:
all.genes.with.pli <- unique(expression[is.ps==T,c("gene.id.x","pLI")])
test.mat <- matrix(c(nrow(ExacPLI[pLI>0.9]),nrow(all.genes.with.pli[pLI>0.9]),nrow(ExacPLI),nrow(all.genes.with.pli)),nrow=2)
fisher.test(test.mat)

##This is to test for difference in Ovary/Testis vs other tissues and add a "*" next to the name in the plots (does not colour red):
tissues<-unique(expression[,tissue])
expression[,mean.log:=log10(mean)]

p.vals<-data.table(tissue=tissues,Ovary=as.double(rep(NA,length(tissues))),Testis=as.double(rep(NA,length(tissues))))
for (germ in c("Ovary","Testis")) {
  for (tis1 in tissues) {
    vals<-c(germ,tis1)
    if (tis1==germ) {
      next;
    }
    test<-expression[tissue %in% vals][is.ps==T]
    ret<-wilcox.test(mean.log ~ tissue, data=test)$p.value
    p.vals[tissue==tis1,(germ):=ret]
  }
}

check.star<-function(tissue.func) {
  x<-p.vals[tissue==tissue.func,Ovary]
  y<-p.vals[tissue==tissue.func,Testis]
  if (x < 1e-3 && y < 1e-3) {
    return(paste(tissue.func, "*", sep=" "))
  } else {
    return(tissue.func)
  }
}
p.vals[,col.name:=mapply(check.star,tissue)]
expression<-merge(expression,p.vals,by="tissue",all.x=T,all.y=F)

print("p.values for comparison between ovary/testis and all other tissues:")
p.vals

##This just checks for difference in expression between pseudogenized vs un-pseudogenized genes within tissues and prints a p-value (all should be very significant)
for (tis in tissues) {
  test<-expression[tissue %in% tis]
  ret<-wilcox.test(mean.log ~ is.ps, data=test)$p.value
  print(paste(tis, ": ", ret, sep = ""))
}

## Look at all tissues seperated into PG T/F
plot<-ggplot(expression,aes(is.ps,mean,colour=is.ps)) + scale_colour_discrete(guide=guide_legend(title="Pseudogene",labels = c("False","True"), breaks=c(T,F))) + geom_boxplot(size=0.5) + facet_wrap(~col.name,nrow=7) + scale_y_log10(name="Mean TPM",labels=c("0.00001",0.1,10,10000),breaks=c(0.00001,0.01,10,10000)) + scale_x_discrete(name = "Is pseudogene?", breaks=c(TRUE,FALSE),labels=c("T","F")) + theme
plot
```

# Supplemental Figure 8

This was generated with a command like the following (the MEI.pli.acc.txt is the same from main text Figure 2):

```
grep intron <MEI>.pli.acc.txt > <MEI>.strand.txt
bcftools query -f '%CHROM\t%POS\t%MEINFO\n' <MEI>.vcf.gz -R <MEI>.strand.txt > <MEI>.strand.2.txt
```

The <MEI>.strand.2.txt file is then matched with the gene annotation to determine Then just match on the gene from which the MEI came and pull the strand information for the gene and calculate proportion and place in the file `RawData/strands.txt`

```{r Strand Figure, fig.height=4, fig.width=7}
strands <- fread("RawData/strands.txt")
setnames(strands,names(strands),c("Strand","prop","ci.lower","ci.upper","total","MEI"))

## Do the chisq.test for each MEI
for (mei in c("ALU","LINE1","SVA")) {

  mat <- matrix(c(strands[MEI==mei & Strand == "same",prop*total],
                  strands[MEI==mei & Strand == "diff",prop*total],
                  strands[MEI==mei & Strand == "same",(total/2)],
                  strands[MEI==mei & Strand == "diff",(total/2)]),
                nrow=2,
                dimnames=list(c("Same Strand","Diff Strand"),c("Actual","Expected")))
                
  print(paste0("p-val for chisq.test for ", mei," : ",chisq.test(mat)$p.value))

}

theme.this.chunk <-theme %+replace% theme(legend.position="right")
plot<-ggplot(strands,aes(MEI,prop,group=Strand,fill=Strand,colour=MEI)) + scale_colour_manual(values=cPalette,labels=c(expression(bolditalic("Alu")),"L1","SVA")) + scale_fill_manual(values=c("#000000","#999999")) + geom_col(position=position_dodge(width=0.85), width=0.85, size=2) + geom_errorbar(aes(ymin=ci.lower,ymax=ci.upper),position=position_dodge(width=0.85), width=0.55, size=2) + scale_y_continuous(expand=c(0,0)) + xlab("") + ylab("Percent of All Sites") + theme.this.chunk
plot
```
